### PR TITLE
Retrieve subjects and temperature data

### PR DESCRIPTION
**Pull request by wgranger**
This first PR includes two simple Azure Functions to retrieve subject and temperature data from a database. You can find more info on the mapping viz tools [wiki](https://github.com/zooniverse/mapping-viz-tools/wiki/What-are-Azure-Functions%3F).

See the README updates here for some info on running this PR.

Currently, this code works by hitting a local postgreSQL database, with the connection string `postgres://postgres:postgres@localhost:5432/sample_db`. Make sure you have postres running with a database titled `sample_db` that contains two tables "subjects" and "temperature".

I have a couple goals for this PR:

1. Make sure the code works, looks clean, and handles errors in a way that makes sense
2. Make sure the code works with a database created from the most recent .csv data posted in the #mapping-viz channel. If all works well, we can upload the database to our Azure instance and try changing the connection string here to hit that database. We should then insert the connection strings and login info here as environment variables, likely in a subsequent PR.

### CODE REVIEW

**Comment by nciemniak** [comment]
might be good to reference these two docs in the readme:
Install Azure CLI: https://docs.microsoft.com/en-us/cli/azure/install-azure-cli-macos?view=azure-cli-latest
Install Azure Functions Core Tools: https://docs.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=macos%2Ccsharp%2Cbash#v2

File: subjects/function.json [hunk 1]
```diff
@@ -0,0 +1,16 @@
+{
+  "bindings": [
+    {
+      "authLevel": "anonymous",
+      "type": "httpTrigger",
+      "direction": "in",
+      "name": "req",
```
Comment (nciemniak) [inline subjects/function.json:7]: ~~what do `req` and `res` refer to? (or what are they short for?)
At first read-through it isnt immediately obvious to me as to what they represent. are they standard binding names used by azure?~~

oh wow, I realized that they're short for request/response immediately after writing out this question. Nevermind :)

File: subjects/index.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    if (!maxLat || !minLat || !maxLon || !minLon) {
+        context.res = {
+            status: 422,
+            body: 'Query params must include maxLat, minLat, maxLon, and minLon'
+        }
+        return context.done()
+    }
+
+    const query = `SELECT * FROM subjects where lat < ${maxLat} and lat > ${minLat}
```
Comment (nciemniak) [inline subjects/index.js]: We dont want to directly append a URL query param value to the SQL query like this, it leaves us open to [SQL injection](https://www.w3schools.com/sql/sql_injection.asp). Instead, use SQL parameters for handling query param inputs. 

See the bit on "Parameterized Queries" here ->
https://node-postgres.com/features/queries
(not 100% sure if youre using node-postgres here but if not, the same principals should apply and its still javascript)

File: temperature/index.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    
+    if (!grid_index) {
+        context.res = {
+            status: 422,
+            body: 'Query param must include grid_index'
+        }
+        return context.done()
+    }
+
+    const query = `SELECT * FROM temperature where sst_grid_index = ${grid_index}`;
```
Comment (nciemniak) [inline temperature/index.js]: Same deal here with SQL Injection/Parameterized Queries, see comment on subject/index.js

File: subjects/index.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+const pg = require('pg');
+
+function executeSQL(context, params) {
+    const connectionString = 'postgres://postgres:postgres@localhost:5432/sample_db'
+    const client = new pg.Client({ connectionString })
+
+    client.connect();
+    const { maxLat, minLat, maxLon, minLon } = params
+
+    if (!maxLat || !minLat || !maxLon || !minLon) {
```
Comment (nciemniak) [inline subjects/index.js:10]: Optional: you could throw this `if` clause into its own function (e.g. `validateParams`) to shrink the size of this main `executeSQL` function. I dont have a strong opinion on this, and it's not like `executeSQL` function is painfully long right now, its just a matter of stylistic preference. If we add any additional validation here though, I would definitely recommend extracting it into its own function.

File: subjects/index.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+            status: 422,
+            body: 'Query params must include maxLat, minLat, maxLon, and minLon'
+        }
+        return context.done()
+    }
+
+    const query = `SELECT * FROM subjects where lat < ${maxLat} and lat > ${minLat}
+        and lon < ${maxLon} and lon > ${minLon}`;
+
+    client.query(query)
```
Comment (nciemniak) [inline subjects/index.js]: Another piece that you could extract into its own function (e.g. `performQuery`) if you want to. I personally would extract it out; it also looks (nearly?) exactly the same as the `client.query` performed in the temperature/index.js file, would be nice to DRY it up.

**Comment by nciemniak** [comment]
Added some comments, @wgranger feel free to rerequest a review once you wanna hand it back to me for another pass over.

Oh oh and almost forgot - maybe we could add some tests? Not sure what tests look like in the land of Azure Functions.

File: subjects/index.js (hunk 4 above)
Comment (wgranger) [reply subjects/index.js:10]: I'll leave this one as is for now, since these are likely the only two functions we'll have for the time being.

**Comment by wgranger** [comment]
@nciemniak Do you think I would need to call `client.end()` after each database connection, GET request? Check [here](https://node-postgres.com/api/client)

**Comment by nciemniak** [comment]
> Do you think I would need to call client.end() after each database connection, GET request?

@wgranger good call - I poked around, that _does_ look like the best practice with handling db connections when using node-postgres. In my experience, there are some tools/packages that do this for you automatically, looks like with node-postgres you need to do so explicitly.

[See this article](https://stackabuse.com/using-postgresql-with-nodejs-and-node-postgres/) - "Note: Don't forget to end() your connection to the client after you've ran the query."

File: package.json [hunk 6]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "mapping-viz-functions",
+  "version": "1.0.0",
+  "description": "",
+  "scripts": {
+    "start": "func start",
+    "test": "jest"
+  },
+  "dependencies": {
+    "jest": "^26.4.2",
```
Comment (wgranger) [inline package.json]: `jest` should be in `devDependencies`

File: performQuery.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+async function performQuery(client, context, query, values) {
```
Comment (nciemniak) [inline performQuery.js]: Lets put this file in a SharedCode or HelperFunctions folder to keep things organized. See this doc: [Azure Functions JavaScript developer guide](https://docs.microsoft.com/en-us/azure/azure-functions/functions-reference-node#folder-structure)

**Review by nciemniak — CHANGES_REQUESTED** [review]
This looks great, just that one bit about moving the performQuery.js file into a separate folder (see comment)

**Review by nciemniak — APPROVED** [review]
Nice! Looks great  🚀 🚀

**Pull request merged**